File: packages/xandr_android/android/src/main/kotlin/de/thekorn/xandr/XandrPlugin.kt
package de.thekorn.xandr

import HostAPIUserId
import XandrHostApi
import android.app.Activity
import com.appnexus.opensdk.ANGDPRSettings
import com.appnexus.opensdk.ANMultiAdRequest
import com.appnexus.opensdk.ANUserId
import com.appnexus.opensdk.ResultCode
import com.appnexus.opensdk.SDKSettings
import com.appnexus.opensdk.XandrAd
import com.appnexus.opensdk.mar.MultiAdRequestListener
import de.thekorn.xandr.listeners.AdInitListener
import de.thekorn.xandr.listeners.XandrInterstitialAdListener
import de.thekorn.xandr.models.FlutterState
import de.thekorn.xandr.models.MultiAdRequestRegistry
import de.thekorn.xandr.models.ads.InterstitialAd
import io.flutter.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import kotlinx.coroutines.ExperimentalCoroutinesApi

class XandrPlugin : FlutterPlugin, ActivityAware, XandrHostApi {
    private lateinit var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding

    private lateinit var flutterState: FlutterState

    // TODO: move to hashmap
    private lateinit var interstitialAd: InterstitialAd

    private lateinit var activity: Activity

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        this.flutterState = FlutterState(binding.applicationContext, binding.binaryMessenger)
        this.flutterState.startListening(this)
        this.flutterPluginBinding = binding
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        this.flutterState.stopListening()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        Log.d(
            "Xandr",
            "--> onAttachedToActivity"
        )
        activity = binding.activity
        this.flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "de.thekorn.xandr/ad_banner",
            BannerViewFactory(
                activity,
                this.flutterState
            )
        )
    }

    override fun onDetachedFromActivityForConfigChanges() {
        Log.d(
            "Xandr",
            "--> onDetachedFromActivityForConfigChanges"
        )
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        Log.d(
            "Xandr",
            "--> onReattachedToActivityForConfigChanges"
        )
        onDetachedFromActivity()
    }

    override fun onDetachedFromActivity() {
        Log.d(
            "Xandr",
            "--> onDetachedFromActivity"
        )
        onDetachedFromActivity()
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    override fun init(memberId: Long, callback: (Result<Boolean>) -> Unit) {
        this.flutterState.memberId = memberId.toInt()
        XandrAd.init(
            memberId.toInt(),
            this.flutterState.applicationContext,
            true,
            true,
            AdInitListener(this.flutterState)
        )
        this.flutterState.isInitialized.invokeOnCompletion {
            callback(Result.success(this.flutterState.isInitialized.getCompleted()))
        }
    }

    override fun loadAd(widgetId: Long, callback: (Result<Boolean>) -> Unit) {
        Log.d("Xandr", "loadAd got called, with widgetId=$widgetId")
        this.flutterState.isInitialized.invokeOnCompletion {
            val ad = this.flutterState.getBannerView(widgetId.toInt())
            Log.d("Xandr", "loadAd found ad for widgetId=$widgetId, ad=$ad")
            ad.loadAd()
            callback(Result.success(true))
        }
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    override fun loadInterstitialAd(
        widgetId: Long,
        placementID: String?,
        inventoryCode: String?,
        customKeywords: Map<String, String>?,
        callback: (Result<Boolean>) -> Unit
    ) {
        interstitialAd = InterstitialAd(activity)
        interstitialAd.adListener = XandrInterstitialAdListener(
            widgetId,
            this.flutterState.flutterApi,
            interstitialAd
        )
        customKeywords?.forEach {
            interstitialAd.addCustomKeywords(it.key, it.value)
        }

        this.flutterState.isInitialized.invokeOnCompletion {
            // / need to make sure the sdk is initialized to access the memberId
            // / docs: Note that if both inventory code and placement ID are passed in, the
            //        inventory code will be passed to the server instead of the placement ID.
            if (inventoryCode != null) {
                interstitialAd.setInventoryCodeAndMemberID(flutterState.memberId, inventoryCode)
            } else {
                interstitialAd.placementID = placementID
            }
            interstitialAd.loadAd()
            Log.d("Xandr.Interstitial", "Loading DONE")
            interstitialAd.isLoaded.invokeOnCompletion {
                callback(Result.success(interstitialAd.isLoaded.getCompleted()))
            }
        }
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    override fun showInterstitialAd(autoDismissDelay: Long?, callback: (Result<Boolean>) -> Unit) {
        if (interstitialAd.isClosed.isCompleted) {
            callback(Result.success(false))
            return
        }

        interstitialAd.isLoaded.invokeOnCompletion {
            if (autoDismissDelay == null) {
                interstitialAd.show()
            } else {
                interstitialAd.showWithAutoDismissDelay(autoDismissDelay.toInt())
            }
            Log.d("Xandr.Interstitial", "show")
        }
        interstitialAd.isClosed.invokeOnCompletion {
            callback(Result.success(interstitialAd.isClosed.getCompleted()))
        }
    }

    override fun setPublisherUserId(publisherUserId: String, callback: (Result<Unit>) -> Unit) {
        SDKSettings.setPublisherUserId(publisherUserId)
        callback(Result.success(Unit))
    }

    override fun getPublisherUserId(callback: (Result<String>) -> Unit) {
        callback(Result.success(SDKSettings.getPublisherUserId()))
    }

    override fun setUserIds(userIds: List<HostAPIUserId>, callback: (Result<Unit>) -> Unit) {
        val uIds = ArrayList<ANUserId>()
        userIds.forEach {
            uIds.add(ANUserId(it.source.toANUserIdSource(), it.userId))
        }
        SDKSettings.setUserIds(uIds)
        callback(Result.success(Unit))
    }

    override fun getUserIds(callback: (Result<List<HostAPIUserId>>) -> Unit) {
        val userIds = SDKSettings.getUserIds()
        val uIds = ArrayList<HostAPIUserId>()
        userIds.forEach {
            uIds.add(it.toHostUserId())
        }
        callback(Result.success(uIds))
    }

    override fun initMultiAdRequest(callback: (Result<String>) -> Unit) {
        val mar = ANMultiAdRequest(
            activity,
            this.flutterState.memberId,
            object : MultiAdRequestListener {
                override fun onMultiAdRequestCompleted() {
                    Log.d("Xandr.MultiAdRequest", "completed")
                }

                override fun onMultiAdRequestFailed(code: ResultCode) {
                    Log.d("Xandr.MultiAdRequest", "failed")
                }
            }
        )
        val id = MultiAdRequestRegistry.initNewRequest(mar)
        callback(Result.success(id))
    }

    override fun disposeMultiAdRequest(multiAdRequestID: String, callback: (Result<Unit>) -> Unit) {
        MultiAdRequestRegistry.removeRequestWithId(multiAdRequestID)
    }

    override fun loadAdsForMultiAdRequest(
        multiAdRequestID: String,
        callback: (Result<Boolean>) -> Unit
    ) {
        callback(Result.success(MultiAdRequestRegistry.load(multiAdRequestID)))
    }

    override fun setGDPRConsentRequired(
        isConsentRequired: Boolean,
        callback: (Result<Unit>) -> Unit
    ) {
        ANGDPRSettings.setConsentRequired(activity, isConsentRequired)
        callback(Result.success(Unit))
    }

    override fun setGDPRConsentString(consentString: String, callback: (Result<Unit>) -> Unit) {
        ANGDPRSettings.setConsentString(activity, consentString)
        callback(Result.success(Unit))
    }

    override fun setGDPRPurposeConsents(purposeConsents: String, callback: (Result<Unit>) -> Unit) {
        ANGDPRSettings.setPurposeConsents(activity, purposeConsents)
        callback(Result.success(Unit))
    }
}

File: packages/xandr_ios/ios/Classes/XandrPlugin.swift
import AppNexusSDK
import Flutter
import UIKit

extension FlutterError: Swift.Error {}

public class XandrPlugin: UIViewController, FlutterPlugin,
  XandrHostApi {
  public static var instance: XandrPlugin?

  public var flutterState: FlutterState?

  private var interstitialAd: InterstitialAd?

  private var marRegistry: MultiAdRequestRegistry = .init()

  private func parentController() -> UIViewController {
    var topController: UIViewController = UIApplication.shared.keyWindow!.rootViewController!
    while topController.presentedViewController != nil {
      topController = topController.presentedViewController!
    }
    return topController
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    // init the plugin and call onRegister
    logger.debug(message: "register plugin instance")
    instance = XandrPlugin()
    instance?.onRegister(registrar: registrar)
  }

  public func onRegister(registrar: FlutterPluginRegistrar) {
    // setup the state, start listening on the host api instance and register the banner factory
    logger.debug(message: "onRegister plugin instance")

    flutterState = FlutterState(binaryMessenger: registrar.messenger())
    flutterState?.startListening(api: self)

    let factory = XandrBannerFactory(
      messenger: registrar.messenger(),
      state: flutterState!
    )
    registrar.register(factory, withId: "de.thekorn.xandr/ad_banner")
  }

  func initXandrSdk(memberId: Int64, publisherId: Int64?, testMode: Bool,
                    completion: @escaping (Result<Bool, Error>) -> Void) {
    // init thge xandr sdk and store the memberId in the state on success
    // return true on success
    logger.debug(message: "initXandrSdk")

    ANSDKSettings.sharedInstance().enableTestMode = testMode

    DispatchQueue.main.async {
      XandrAd.sharedInstance()
        .initWithMemberID(Int(memberId), preCacheRequestObjects: true) { [self]
          success in
            if success {
              logger.debug(message: "initXandrSdk was successfull")
              flutterState?.memberId = Int(memberId)
              if publisherId != nil {
                flutterState?.publisherId = Int(publisherId!)
              }
              flutterState?.setIsInitialized(success: true)
              completion(.success(true))
            } else {
              logger.debug(message: "initXandrSdk failed")
              flutterState?.memberId = Int(memberId)
              if publisherId != nil {
                flutterState?.publisherId = Int(publisherId!)
              }
              flutterState?.setIsInitialized(success: false)
              completion(.failure(NSError(
                domain: "XandrAd.sharedInstance().initWithMemberID",
                code: 1,
                userInfo: nil
              )))
            }
        }
    }
  }

  func loadAd(widgetId: Int64, completion: @escaping (Result<Bool, any Error>) -> Void) {
    logger.debug(message: "load Ad for widgetId=\(widgetId)")

    flutterState?.setIsInitializedCompletionHandler { [self] result in
      do {
        let ad = try flutterState?.getXandrBanner(id: widgetId)
        logger
          .debug(
            message: "loadAd found ad for widgetId=\(widgetId), ad=\(String(describing: ad))"
          )
        ad?.loadAd()

        completion(.success(true))
      } catch let err {
        logger.error(message: "Loading ad for widgetId=\(widgetId) results in error=\(err)")
        completion(.success(false))
      }
    }
  }

  func loadInterstitialAd(widgetId: Int64, placementID: String?, inventoryCode: String?,
                          customKeywords: [String: [String]]?,
                          completion: @escaping (Result<Bool, Error>) -> Void) {
    interstitialAd = InterstitialAd()
    interstitialAd!.delegate = self

    customKeywords?.forEach { keyword in
      for value in keyword.value {
        interstitialAd?.addCustomKeyword(withKey: keyword.key, value: value)
      }
    }

    flutterState?.setIsInitializedCompletionHandler(handler: { _ in
      if self.flutterState?.memberId != nil {
        self.interstitialAd?.setInventoryCode(inventoryCode, memberId: self.flutterState!.memberId)
      } else if placementID != nil {
        self.interstitialAd?.placementId = placementID
      }
      self.interstitialAd?.load()
      logger.debug(message: "Xandr.Interstitial Loading DONE")

      self.interstitialAd?.isLoaded.invokeOnCompletion { _ in
        completion(Result.success(self.interstitialAd!.isLoaded.getCompleted()!))
      }
    })
  }

  func showInterstitialAd(autoDismissDelay: Int64?,
                          completion: @escaping (Result<Bool, Error>) -> Void) {
    if interstitialAd?.isClosed.isCompleted ?? false {
      completion(Result.success(false))
      return
    }

    if !(interstitialAd?.isReady ?? false) {
      completion(Result.success(false))
      return
    }

    interstitialAd?.isLoaded.invokeOnCompletion { _ in
      if autoDismissDelay == nil {
        self.interstitialAd?.display(from: self.parentController())
      } else {
        self.interstitialAd?.display(
          from: self.parentController(),
          autoDismissDelay: Double(autoDismissDelay!)
        )
      }
    }

    interstitialAd?.isClosed.invokeOnCompletion { _ in
      completion(Result.success(self.interstitialAd!.isClosed.getCompleted()!))
    }
  }

  func initMultiAdRequest(completion: @escaping (Result<String, Error>) -> Void) {
    guard let memberId = flutterState?.memberId else {
      completion(Result.failure(XandrPluginError.noMemberId))
      return
    }
    let mar: ANMultiAdRequest?
    if flutterState?.publisherId != nil {
      mar = ANMultiAdRequest(
        memberId: memberId,
        publisherId: (flutterState?.publisherId)!,
        andDelegate: self
      )
    } else {
      mar = ANMultiAdRequest(memberId: memberId, andDelegate: self)
    }
    if mar == nil {
      completion(Result.failure(XandrPluginError.noMemberId))
      return
    }
    let id = marRegistry.initNewRequest(mar!)
    completion(Result.success(id))
  }

  func disposeMultiAdRequest(multiAdRequestID: String,
                             completion: @escaping (Result<Bool, Error>) -> Void) {
    let result = marRegistry.removeRequestWithId(multiAdRequestID)
    completion(Result.success(result))
  }

  func loadAdsForMultiAdRequest(multiAdRequestID: String,
                                completion: @escaping (Result<Bool, Error>) -> Void) {
    let result = marRegistry.load(multiAdRequestID)
    completion(Result.success(result))
  }

  func setPublisherUserId(publisherUserId: String,
                          completion: @escaping (Result<Bool, Error>) -> Void) {
    ANSDKSettings.sharedInstance().publisherUserId = publisherUserId
    completion(Result.success(true))
  }

  func getPublisherUserId(completion: @escaping (Result<String, Error>) -> Void) {
    completion(Result.success(ANSDKSettings.sharedInstance().publisherUserId ?? ""))
  }

  func setUserIds(userIds: [HostAPIUserId],
                  completion: @escaping (Result<Bool, Error>) -> Void) {
    var uIds: [ANUserId] = []
    for it in userIds {
      let newId = ANUserId(anUserIdSource: it.source.toANUserIdSource(), userId: it.userId)
      if newId != nil {
        uIds.append(newId!)
      }
    }
    ANSDKSettings.sharedInstance().userIdArray = uIds
    completion(Result.success(true))
  }

  func getUserIds(completion: @escaping (Result<[HostAPIUserId], Error>) -> Void) {
    var userIds: [HostAPIUserId] = []
    ANSDKSettings.sharedInstance().userIdArray?.forEach { it in
      do {
        let newUserIds = try HostAPIUserId(
          source: it.source.toHostAPIUserIdSource(),
          userId: it.userId
        )
        userIds.append(newUserIds)
      } catch {
        logger.debug(message: "no valid user source")
      }
    }
    completion(Result.success(userIds))
  }

  func setGDPRConsentRequired(isConsentRequired: Bool,
                              completion: @escaping (Result<Bool, Error>) -> Void) {
    ANGDPRSettings.setConsentRequired(isConsentRequired ? 1 : 0)
    completion(Result.success(true))
  }

  func setGDPRConsentString(consentString: String,
                            completion: @escaping (Result<Bool, Error>) -> Void) {
    ANGDPRSettings.setConsentString(consentString)
    completion(Result.success(true))
  }

  func setGDPRPurposeConsents(purposeConsents: String,
                              completion: @escaping (Result<Bool, Error>) -> Void) {
    ANGDPRSettings.setPurposeConsents(purposeConsents)
    completion(Result.success(true))
  }

  func setAutoRefreshInterval(autoRefreshIntervalInSeconds: Int64, inventoryCode: String?,
                              placementID: String?,
                              completion: @escaping (Result<Bool, Error>) -> Void) {
    do {
      let ad = try flutterState?.getXandrBannerWithCode(
        inventoryCode: inventoryCode,
        placementID: placementID
      )
      logger
        .debug(
          message: "setAutoRefreshInterval found ad for inventoryCode=\(String(describing: inventoryCode)), placementID=\(String(describing: placementID)), ad=\(String(describing: ad))"
        )
      ad?.banner?.autoRefreshInterval = Double(autoRefreshIntervalInSeconds)
      completion(.success(true))
    } catch let err {
      logger
        .error(
          message: "Loading ad for inventoryCode=\(String(describing: inventoryCode)), placementID=\(String(describing: placementID)) results in error=\(err)"
        )
      completion(.success(false))
    }
    completion(Result.success(true))
  }
}

extension HostAPIUserIdSource {
  func toANUserIdSource() -> ANUserIdSource {
    switch self {
    case .criteo: ANUserIdSource.criteo
    case .theTradeDesk:
      ANUserIdSource.theTradeDesk
    case .netId:
      ANUserIdSource.netId
    case .liveramp:
      ANUserIdSource.liveRamp
    case .uid2:
      ANUserIdSource.UID2
    }
  }
}

extension String {
  func toHostAPIUserIdSource() throws -> HostAPIUserIdSource {
    switch self {
    case "criteo.com":
      HostAPIUserIdSource.criteo
    case "uidapi.com":
      HostAPIUserIdSource.uid2
    case "netid.de":
      HostAPIUserIdSource.netId
    case "liveramp.com":
      HostAPIUserIdSource.liveramp
    case "adserver.org":
      HostAPIUserIdSource.liveramp
    default:
      throw XandrPluginError.notValidSource
    }
  }
}

// ANInterstitialAdDelegate
extension XandrPlugin: ANInterstitialAdDelegate {
  public func adDidReceiveAd(_ ad: Any) {
    interstitialAd?.isLoaded.complete(true)
  }

  public func adDidClose(_ ad: Any) {
    interstitialAd?.isClosed.complete(true)
  }
}

// ANMultiAdRequestDelegate
extension XandrPlugin: ANMultiAdRequestDelegate {
  public func multiAdRequestDidComplete(_ mar: ANMultiAdRequest) {
    logger.debug(message: "Xandr.MultiAdRequest completed")
  }

  public func multiAdRequestDidFailWithError(_ error: NSError) {
    logger.debug(message: "Xandr.MultiAdRequest failed")
  }
}
